### DISCUSSION TITLE
Elegant way to do headless validation?

### DISCUSSION THREAD

**Discussion by fappaz** [react-hook-form/react-hook-form#7999] [Q&A] [answered]
Imagine a review page showing read-only values, but the user cannot proceed to the next page while there are [validation errors](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input#client-side_validation).

One of the options I came up with was to add hidden inputs, then programmatically click on a hidden submit button as soon as the page loads (sample code can be found on [codesandbox](https://codesandbox.io/s/rhf-headless-validation-kql85g?file=/src/App.js)). Any invalid values would be highlighted (e.g.: I can ask the user to edit them).

However, this approach seem way too "hacky" for my taste. Would anybody know a more elegant way to achieve this?

**Comment by Moshyfawn** [comment, marked as answer]
I don't completely understand your use-case, but to validate your form on first render, you can do
```js
useEffect(() => {
  trigger()
}, [trigger])
```

**Reply by fappaz** [reply]
Nice, I wasn't aware of the trigger function, it indeed allows me to validate the form fields without simulating a submit button.

However, it seems I would still need to insert input elements into the DOM to be able to trigger the validation. Is there a way I can set up the fields headlessly?

Just to clarify my use case, I need to display read-only data (e.g.: name, email, etc) coming from a 3rd party API, and highlight all fields that don't meet the validation rules (e.g.: name must be longer than N chars, email must contain "@", etc) coming from another API. 

Ideally I want to reuse react-hook-form's built-in validation, without displaying the inputs on my web page. The sample code provided is hiding the elements, but that doesn't seem good practice. Any thoughts?

**Reply by Moshyfawn** [reply]
You can use `Controller` / `useController` without the input element.
Since you need that validation piece, you want the `rules` prop of `Controller` to set the validation rules tied to the field's name. Without the validation requirement you could just use the `watch` API.

I've created [a very simple CSB](https://codesandbox.io/s/elastic-tu-ubx971?file=/src/App.tsx) to display the example of how I would approach the task. I didn't include the actual fields config, but you can get the main idea and tweak it to your needs.

**Reply by fappaz** [reply]
Cool, that's more what I was looking for, thanks @Moshyfawn
